### PR TITLE
start a blog post draft on the optimization rule dsl

### PR DESCRIPTION
**Pull request by cfbolz**

### CODE REVIEW

**Review by tekknolagi — APPROVED** [review]
Love this. Publish it! Nits only

File: posts/2024/10/jit-peephole-dsl.rst [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+we express these rewrites up to now is very imperative and verbose. Here's a
+snippet of RPython code that shows some rewrites for integer multiplication. You
+don't need to understand it, but basically it's in very imperative style and
+there's quite a lot of boilerplate.
+
+.. __: /posts/2024/07/mining-jit-traces-missing-optimizations-z3.html
+
+.. code:: python
+
+    def optimize_INT_MUL(self, op):
```
[the remark this answers is not in the document]
Comment (cfbolz) [reply posts/2024/10/jit-peephole-dsl.rst:45]: I'll add comments to the code that explain which rewrite is being done.

File: posts/2024/10/jit-peephole-dsl.rst [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+                        op = self.replace_op_with(
+                                op, rop.INT_LSHIFT, args=[rhs, shiftvar])
+                        self.optimizer.send_extra_operation(op)
+                        return
+            return self.emit(op)
+
+Adding more rules to these functions was very tedious and got super confusing
+after a while. In addition I was worried about making mistakes when writing this
+kind of code, and I had basically no feedback at all about which of these rules
+are actually applied a lot in real programs.
```
[the remark this answers is not in the document]
Comment (cfbolz) [reply posts/2024/10/jit-peephole-dsl.rst:90]: oh, I forgot. what did we discuss on the boat that was relevant?

File: posts/2024/10/jit-peephole-dsl.rst [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+
+    add_zero: int_add(x, 0)
+        => x
+
+The name of the rule is ``add_zero``. It matches operations in the trace of the
+form ``int_add(x, 0)``, where ``x`` will match anything and ``0`` will match only the
+constant zero. After the ``=>`` arrow is the target of the rewrite, i.e. what the
+operation is rewritten to, in this case ``x``.
+
+The rule language knowns which of the operations are commutative, so ``add_zero``
```
[the remark this answers is not in the document]
Comment (cfbolz) [reply posts/2024/10/jit-peephole-dsl.rst:117]: there's simply a list of commutative versions ;-). I added that.

about nesting, yeah, all the commutative combinations of the operation names that appear in the head of the rule are tried. not sure where to write that though.

File: posts/2024/10/jit-peephole-dsl.rst [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+Here's a rule with a more complicated pattern::
+
+    sub_add: int_sub(int_add(x, y), y)
+        => x
+
+This pattern matches ``int_sub`` operations, where the first argument was
+produced by an ``int_add`` operation. In addition, one of the arguments of the
+addition has to be the same as the second argument of the subtraction.
+
+The constants ``MININT``, ``MAXINT`` and ``LONG_BIT`` (which is either 32 or 64) can
```
[the remark this answers is not in the document]
Comment (cfbolz) [reply posts/2024/10/jit-peephole-dsl.rst:137]: no, it creates a version of the rule that depends on which platform you're on. added a clarification.

File: posts/2024/10/jit-peephole-dsl.rst [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+because the former two rules optimize to a constant and a variable
+respectively, while the latter produces a new operation as the result.
+
+The rule ``sub_add_consts`` has a possible problem, which is that if the
+intermediate result of the ``int_add`` operation in the rule head is used by
+some other operations, then the ``sub_add_consts`` rule does not actually
+reduce the number of operations (and might actually make things slightly worse
+due to increased register pressure). However, currently it would be extremely
+hard to take that kind of information into account in the optimization pass of
+the JIT, so we optimistically apply the rules anyway.
```
[the remark this answers is not in the document]
Comment (cfbolz) [reply posts/2024/10/jit-peephole-dsl.rst:233]: we do have to be careful to write monotonic rules. let me think about where to add that.

File: posts/2024/10/jit-peephole-dsl.rst [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+  started a project with Nico Rittinhaus to try to improve the situation a bit
+  more generally.
+
+- A more advanced direction of work would be to implement a simplified form of
+  `e-graphs`__ (or `ae-graphs`__). The JIT has like half of an e-graph data
+  structure already, and we probably can't afford a full one in terms of compile
+  time costs, but maybe we can have two thirds or something?
+
+- The DSL isn't integrated with common subexpression elimination yet, and that
+  could potentially yield some further improvement.
```
[the remark this answers is not in the document]
Comment (cfbolz) [reply posts/2024/10/jit-peephole-dsl.rst:448]: I've removed it. I have various vague ideas but they are probably like e-graphs anyway, so no reason to mention them twice.

File: posts/2024/10/jit-peephole-dsl.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (tekknolagi) [reply posts/2024/10/jit-peephole-dsl.rst]: Something John was working on that inspired this whole z3 adventure

**Comment by tekknolagi** [comment]
Great stuff. Looking forward to seeing it live

**Pull request merged**